src/aosm/azext_aosm/common/registry.py

```diff
@@ -0,0 +1,469 @@
+            logger.warning(
+                (
+                    "Failed to copy %s from %s to %s. If this failure is because it already "
+                    "exists in the target registry we can continue. If the failure was for "
+                    "another reason, for example it does not exist in the source registry, "
+                    "this is likely fatal. Attempting to continue.\n"
+                    "%s"
+                ),
+                source_image,
+                self.registry_name,
+                target_acr,
+                error,
+            )
+
+    def find_image(
+        self, image: str, version: str
+    ) -> Union[Tuple["AzureContainerRegistry", str], Tuple[None, None]]:
+        """
+        Find whether the given image exists in this registry.
+
+        :param image: The image to find in the registry
+            (the image should be stripped of all namespace)
+        :param version: The version of the image (also known as image tag)
+        :return: The registry and namespace for the image
+        """
+        for namespace in self.registry_namespaces:
+            image_with_namespace = f"{namespace}{image}:{version}"
+            try:
+                acr_source_get_images_cmd = [
+                    str(shutil.which("az")),
+                    "acr",
+                    "repository",
+                    "show",
+                    "--name",
+                    self.registry_name,
+                    "--image",
+                    image_with_namespace,
+                ]
+
+                call_subprocess_raise_output(acr_source_get_images_cmd)
+
+                return (self, namespace)
+
+            except CLIError as error:
+                logger.debug(
+                    ("Image %s, version %s not found in %s registry."),
+                    image_with_namespace,
+                    version,
+                    self.registry_name,
+                )
+                logger.debug(error, exc_info=True)
+
+        return None, None
+
+
+class ContainerRegistryHandler:
+    """
+    A class to handle all the registries and their images.
+    """
+
+    def __init__(self, image_sources):
+        self.image_sources = image_sources
+        self.registry_list = self._create_registry_list()
+
+    def _create_registry_list(self) -> List[ContainerRegistry]:
+        """
+        Create a list of registry objects from the registries provided by the user.
+
+        :return: A list of registry objects
+        """
+
+        registry_name_to_class_dict: Dict[str, ContainerRegistry] = {}
+        registry_object_list: List[ContainerRegistry] = []
+
+        logger.debug("Creating registry list from the provided registries")
+
+        for registry in self.image_sources:
+            registry = clean_registry_name(registry)
+
+            # If registry matches the pattern of myacr1.azurecr.io/**,
+            # then create a an instance of the ACR registry class
```

Comment:
This comment is out of place, but I'd argue isn't needed as the code is clear enough, so I'd just delete it.